pkg/controller/endpointslice/utils.go

```diff
@@ -60,7 +62,11 @@ func podToEndpoint(pod *corev1.Pod, node *corev1.Node, service *corev1.Service,
 		}
 	}
 
-	ready := service.Spec.PublishNotReadyAddresses || podutil.IsPodReady(pod)
+	accepting := podutil.IsPodReady(pod)
+	terminating := pod.DeletionTimestamp != nil
+	// For compatibility reasons, "ready" should never be "true" if a pod is terminatng, unless
+	// publishNotReadyAddresses was set.
+	ready := service.Spec.PublishNotReadyAddresses || (accepting && !terminating)
```

Comment:
It seems like we need to call out this difference in the docs for the accepting condition. I generally agree with dropping the `publishNotReadyAddresses` check here for `accepting`, but that may be unexpected for some.

Reply:
> I generally agree with dropping the publishNotReadyAddresses check here for accepting, but that may be unexpected for some.

I was wondering about this too. When this gate goes beta and on by default, `publishNotReadyAddresses` wouldn't really apply anymore (at least with EndpointSlice) since all endpoints are included but with the new conditions. Is this technically a breaking change? I'm in favor of just dropping `publishNotReadyAddresses` in the context of EndpointSlice but I don't know if that's okay.

Reply:
It looks like the definition of `publishNotReadyAddresses` has [evolved a bit](https://github.com/kubernetes/kubernetes/pull/90783) since the last time I looked at it. Given the recent clarifications here, I think we probably need to continue to support this both for `ready` and `serving`.

Reply:
Ah, thanks for digging that up.